python/cudf/cudf/_libxx/table.pyx

```diff
@@ -66,39 +66,45 @@ cdef class Table:
     cdef Table from_unique_ptr(
         unique_ptr[table] c_tbl,
         object column_names,
-        object index_names=None
+        object index_names=[],
+        object index_pos=None
     ):
         """
         Construct a Table from a unique_ptr to a cudf::table.
-
         Parameters
         ----------
         c_tbl : unique_ptr[cudf::table]
         index_names : iterable
         column_names : iterable
+        index_pos : iterable, column indices corresponding to column_names
         """
+
         cdef vector[unique_ptr[column]] columns
         columns = c_tbl.get()[0].release()
 
         cdef vector[unique_ptr[column]].iterator it = columns.begin()
 
-        # First construct the index, if any
+        num_cols_input = 0
+        num_cols_input += len(column_names) if column_names is not None else 0
+        num_cols_input += len(index_names) if index_names is not None else 0
+
+        cols_py = OrderedDict()
+        for i in range(num_cols_input):
+            cols_py[i] =(Column.from_unique_ptr(
+                move(dereference(it))
+            ))
+            it += 1
+
         index = None
-        if index_names is not None:
+        if index_names not in [None, []]:
             index_columns = []
-            for _ in index_names:
-                index_columns.append(Column.from_unique_ptr(
-                    move(dereference(it))
-                ))
-                it += 1
+            if index_pos is None:
+                index_pos = range(len(index_names))
+            for idx in index_pos:
+                index_columns.append(cols_py.pop(idx))
             index = Table(OrderedColumnDict(zip(index_names, index_columns)))
 
-        # Construct the data OrderedColumnDict
-        data_columns = []
-        for _ in column_names:
-            data_columns.append(Column.from_unique_ptr(move(dereference(it))))
-            it += 1
-        data = OrderedColumnDict(zip(column_names, data_columns))
+        data = OrderedColumnDict(zip(column_names, cols_py.values()))
```

Comment:
If you're not grabbing every column from the table this now has a size mismatch because we're just dumping what's left of the OrderedDict into here while `column_names` can be arbitrarily small.

Reply:
Correct, we should probably just iterate through the length of `column_names` instead.